contracts/wrapped-assets/shiba/WrappedShibaSwapLp.sol

```diff
@@ -0,0 +1,414 @@
+// SPDX-License-Identifier: bsl-1.1
+
+/*
+  Copyright 2021 Unit Protocol: Artem Zakharov (az@unit.xyz).
+*/
+pragma solidity 0.7.6;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+import "../../helpers/ReentrancyGuard.sol";
+import "../../helpers/TransferHelper.sol";
+import "../../Auth2.sol";
+import "../../interfaces/IVault.sol";
+import "../../interfaces/IERC20WithOptional.sol";
+import "../../interfaces/wrapped-assets/IWrappedAsset.sol";
+import "../../interfaces/wrapped-assets/ITopDog.sol";
+import "../../interfaces/wrapped-assets/ISushiSwapLpToken.sol";
+
+/**
+ * @title ShibaSwapWrappedLp
+ **/
+contract WrappedShibaSwapLp is IWrappedAsset, Auth2, ERC20, ReentrancyGuard {
+    using SafeMath for uint256;
+
+    uint256 public constant MULTIPLIER = 1e12;
+
+    IVault public immutable vault;
+    ITopDog public immutable topDog;
+    uint256 public immutable topDogPoolId;
+    IERC20 public immutable boneToken;
+
+    uint256 public lastKnownBonesBalance;
+    uint256 public accBonePerShare; // Accumulated BONEs per share, times MULTIPLIER. See below.
+
+    uint256 public feePercent = 10;
+    uint256 public constant FEE_DENOMINATOR = 100;
+    address public feeReceiver;
+
+    // Reward debt. See explanation below.
+    // similar to TopDog contract https://etherscan.io/address/0x94235659cf8b805b2c658f9ea2d6d6ddbb17c8d7
+    // but we update accBonePerShare every transaction
+    //
+    // We do some fancy math here. Basically, any point in time, the amount of BONEs
+    // entitled to a user but is pending to be distributed is:
+    //
+    //   pending reward = (totalBalanceOf(user) * accBonePerShare) - rewardDebt[user]
+    //
+    // Whenever a user deposits or withdraws LP tokens to a pool. Here's what happens:
+    //   1. The pool's `accBonePerShare` gets updated.
+    //   2. User receives the pending reward sent to his/her address.
+    //   3. User's `totalBalanceOf` gets updated (since minted/burned tokens for user)
+    //   4. User's `rewardDebt` gets updated.
+    //
+    // on transfers balanceOf + deposited to vault = const
```

Comment:
The impl is quite noisy. E.g., `_sendPendingRewardInternal` calls and `rewardDebts` updates here and there. 
Could we borrow the idea from https://github.com/unitprotocol/fee-distribution/blob/main/contracts/USDPStaking.sol (that is a more wide-spread approach) ?

Reply:
(btw, it makes sense to generalize that code as I can't endure more copy-paste reviewing)

Reply:
didn't do base contract for reward distribution since another way was found. Next time